internal/v2/s2av2.go

```diff
@@ -0,0 +1,99 @@
+// Package s2av2 provides the S2Av2 transport credentials used by a gRPC
+// application.
+package s2av2
+
+import (
+	"errors"
+	"context"
+	"net"
+	"crypto/tls"
+
+	"google.golang.org/grpc/credentials"
+	"github.com/google/s2a-go/internal/v2/tls_config_store"
+)
+
+const (
+	s2aSecurityProtocol = "s2av2"
+)
+
+type s2av2TransportCreds struct {
+	info     *credentials.ProtocolInfo
+	isClient bool
+	serverName string
+}
+
+// NewClientCreds returns a client-side transport credentials object that uses
+// the S2Av2 to establish a secure connection with a server.
+func NewClientCreds(serverName string) (credentials.TransportCredentials, error) {
+	creds := &s2av2TransportCreds{
+		info: &credentials.ProtocolInfo{
+			SecurityProtocol: s2aSecurityProtocol,
+		},
+		isClient: true,
+		serverName: serverName,
+	}
+	return creds, nil
+}
+
+// NewServerCreds returns a server-side transport credentials object that uses
+// the S2Av2 to establish a secure connection with a client.
+func NewServerCreds() (credentials.TransportCredentials, error) {
+	creds := &s2av2TransportCreds{
+		info: &credentials.ProtocolInfo{
+			SecurityProtocol: s2aSecurityProtocol,
+		},
+		isClient: false,
+	}
+	return creds, nil
+}
+
+// ClientHandshake performs a client-side mTLS handshake using the S2Av2.
+func (c *s2av2TransportCreds) ClientHandshake(ctx context.Context, serverAuthority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	if !c.isClient {
+		return nil, nil, errors.New("client handshake called using server transport credentials")
+	}
+	// TODO(rmehta19): Create a stub to S2Av2.
+	var config *tls.Config
+	if c.serverName == "" {
+		config = tlsconfigstore.GetTlsConfigurationForClient(serverAuthority)
```

Comment:
Forgot about removing the port using net.SplithostPort, should that be done to serverAuthority here before passing to GetTlsConfigurationForClient?

Reply:
I don't want to block this PR any further so let's do this in a follow-up. Can we leave a TODO?

Note that we want to remove the port from `serverAuthority` and from `c.serverName`. We can maybe remove the port from `c.serverName` when it's set in OverrideServerName.

Reply:
Added TODO to do this in follow up PR.